rust/lance-encoding/src/encoder.rs

```diff
@@ -443,10 +565,24 @@ impl ArrayEncodingStrategy for CoreArrayEncodingStrategy {
         let data_type = arrays[0].data_type();
         let use_dict_encoding = data_type == &DataType::Utf8
             && check_dict_encoding(arrays, get_dict_encoding_threshold());
+
+        let binary_datatypes = [
+            DataType::Utf8,
+            DataType::LargeUtf8,
+            DataType::Binary,
+            DataType::LargeBinary,
+        ];
+
+        let byte_width = &mut 0;
+        let use_fixed_size_encoding =
+            binary_datatypes.contains(data_type) && check_fixed_size_encoding(arrays, byte_width);
```

Comment:
It's ok to use the arrays to check if an encoding is applicable.  We need to do that for both dictionary and here for fixed size string.  These are unlike the other encodings where we just use the type.

The way it's coded up though requires passing these `use_xyz_encoding` args to `array_encoder_from_type` which is a little awkward.  I could envision two approaches.

First approach.  If we've determined that these encodings are correct, why even bother calling `array_encoder_from_type`?  We should be able to just new up the encoding and return it.

Second approach, if some encodings need the arrays (and not just the data type) then we could change `array_encoder_from_type` to `choose_array_encoder` and pass in the arrays.  Something like...

```
// Gather common info needed by all encodings
let data_size = arrays
    .iter()
    .map(|arr| arr.get_buffer_memory_size() as u64)
    .sum::<u64>();
let data_type = arrays[0].data_type();
// Pick an encoding
Self::choose_array_encoder(data_type, data_size, arrays, self.version, Some(&field.metadata))
```

In other words, we rename `array_encoder_from_type` as `choose_array_encoder`.  Then the "does dictionary apply" or "does fixed size apply" can go into the same method as all the other "does xyz encoder apply" logic.

Reply:
I don't think we can use this approach directly
For example, the `use_dict_encoding` flag needs to be computed for arrays with a binary datatype.
And now for the inner bytes we need to figure out which encoding to use (fixed size binary, binary). This requires calling `choose_array_encoder` again. In this case, if we don't pass this parameter recursively, we may end up using the dict encoding again and get a stack overflow.

Reply:
We don't compute the actual dictionary in the beginning, so `arrays` will still be the parent array and can have duplicates as per the current structure
For now, I have restructured the `use_fixed_size_encoding()` check and moved it inside `choose_array_encoder`. 
I agree this "choose encoder" workflow probably needs more work in a follow-up.